src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs

```diff
@@ -131,49 +141,180 @@ protected override void DetachCore()
             {
-                Task res = this.ProcessSendCmpctPayload(node, sendCmpctPayload);
-                this.logger.LogTrace("(-)[SEND_CMPCT_PAYLOAD]");
-                return res;
+                forkPoint = this.chain.FindFork(getBlocksPayload.BlockLocators.Blocks);
+                if (forkPoint == null)
+                {
+                    this.logger.LogTrace("(-)[NO_FORK_POINT]");
+                    return;
+                }
+
+                // In case of reorg, we just try again, eventually we succeed.
+                if (chainTip.FindAncestorOrSelf(forkPoint) == null)
+                {
+                    chainTip = this.chain.Tip;
+                    forkPoint = null;
+                }
             }
 
-            var sendHeadersPayload = message.Message.Payload as SendHeadersPayload;
-            if (sendHeadersPayload != null)
-                this.PreferHeaders = true;
+            this.logger.LogTrace("Fork point is '{0}'.", forkPoint);
+
+            // If block store is lower than the fork point, or it is on different chain, we don't have anything to contribute to this peer at this point.
+            if (blockStoreTip.FindAncestorOrSelf(forkPoint) == null)
+            {
+                this.logger.LogTrace("(-)[FORK_OUTSIDE_STORE]");
+                return;
+            }
+
+            // Now we compile a list of blocks we want to send to the peer as inventory vectors.
+            // This is needed as we want to traverse the chain in forward direction.
+            int maxHeight = Math.Min(blockStoreTip.Height, forkPoint.Height + InvPayload.MaxGetBlocksInventorySize);
+            ChainedBlock lastBlock = blockStoreTip.GetAncestor(maxHeight);
+            int headersCount = maxHeight - forkPoint.Height;
+
+            this.logger.LogTrace("Last block to announce is '{0}', number of blocks to announce is {1}.", lastBlock, headersCount);
+
+            ChainedBlock[] headersToAnnounce = new ChainedBlock[headersCount];
+            for (int i = headersCount - 1; i >= 0; i--)
+            {
+                headersToAnnounce[i] = lastBlock;
+                lastBlock = lastBlock.Previous;
+            }
+
+            // Now we compile inventory payload and we also consider hash stop given by the peer
+            // and we also check if the blocks to be announced are really present in the store.
+            bool sendContinuation = true;
+            ChainedBlock lastAddedChainedBlock = null;
+            var inv = new InvPayload();
+            for (int i = 0; i < headersToAnnounce.Length; i++)
+            {
+                ChainedBlock chainedBlock = headersToAnnounce[i];
+                if (chainedBlock.HashBlock == getBlocksPayload.HashStop)
+                {
+                    this.logger.LogTrace("Hash stop has been reached.");
+                    break;
+                }
+
+                if (!await this.blockRepository.ExistAsync(chainedBlock.HashBlock).ConfigureAwait(false))
```

Comment:
This step is probably not necessary, we could not get here if the block was not in store right?
This line earlier means the store and all its previous blocks are in place.
```
 ChainedBlock blockStoreTip = this.chain.GetBlock(this.blockRepository.BlockHash);
```

Reply:
it seems that you are right, i can't see how a block could NOT YET be in store at this point

it can be out of the store due to reorg, in which case it would make sense to stop the loop, however, that can happen even after that, so it makes no sense to check for that (again)

i'll remove it